PR: Create new API for plugins and widgets

goanpeca opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)
<!--- Remember that an image/animation is worth a thousand words! --->
* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))


<!--- Explain what you've done and why --->




### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #12180


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @goanpeca 

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @goanpeca! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`spyder/api/completion.py`](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/completion.py):

> [Line 43:3](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/completion.py#L43): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four

* In the file [`spyder/api/plugins.py`](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py):

> [Line 319:1](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L319): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 516:30](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L516): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 517:30](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L517): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 518:30](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L518): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 521:49](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L521): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 608:80](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L608): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 677:47](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L677): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 784:78](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L784): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 898:5](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L898): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 900:5](https://github.com/spyder-ide/spyder/blob/69624b612475e1518a257dc34cd888dd140a0777/spyder/api/plugins.py#L900): [E301](https://duckduckgo.com/?q=pep8%20E301) expected 1 blank line, found 0

goanpeca commented:
@ccordoba12 besides fixing the errors there are some issues that need to be settled on this API and that is what is common for both types of plugins.

I made the following changes:
- `register_plugin` is needed on both
- `on_first_registration` should be available on both
- `get_plugin_title` and `get_plugin_icon`should be available on both since a SpyderPlugin can defne configuration and it also needs a localized name.
- `register_shortcut` needs to be also available on both (autopep8 plugin)

goanpeca commented:
@ccordoba12 please take a look at the `FIXME`comments I left on different methods and let me know what you think

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -8,69 +8,165 @@
 
-Here, 'plugins' are Qt widgets that can make changes to Spyder's
-main window and call other plugins directly.
+Here, 'plugins' are Qt widgets that can make changes to Spyder's main window
```
```suggestion
Here, 'plugins' are Qt objects that can make changes to Spyder's main window
```

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -8,69 +8,165 @@
+Here, 'plugins' are Qt widgets that can make changes to Spyder's main window
+and call other plugins directly.
+
+There are two types of plugins available.
```
```suggestion
There are two types of plugins available:
```

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -8,69 +8,165 @@
+
+There are two types of plugins available.
+* SpyderPlugin is a plugin that does not create a new dock/pane on Spyder
+  main window
```
```suggestion
  main window.
```

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -8,69 +8,165 @@
+    DESCRIPTION = None
+
+    # TODO: add a REQUIRES list to hold names of plugins that this plugin
+    # needs, this will allow to define the order of plugin registry on startup
```
```suggestion
    # TODO: Add a REQUIRES list to hold names of plugins that this plugin
    # needs. That will allow to define the order of plugin registry at startup.
```

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -8,69 +8,165 @@
+        messageBox.setStandardButtons(QMessageBox.Ok)
+        messageBox.show()
+
+    # TODO: This should be public? since it is used on mainwindow.py
```
If it's not used by spyder-notebook, spyder-terminal, etc, directly, we should leave it as private.

goanpeca replied:
This was an old comment I forgot to erase ;-)

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -89,25 +185,27 @@ def set_option(self, option, value, section=None):
+        if self.main is not None:
+            self.main.statusBar().showMessage(message, timeout)
 
+    # FIXME: Should be start_
```
This name makes sense because the plugin is not starting a long process here. It's only informing the main window that it's starting one.

goanpeca replied:
the `-ing`form of verbs feels odd in naming. Since this methods relate to actions, it makes more sense to use `on_`, `before_, `after_`

ccordoba12 replied:
Ok, `before_long_process` sounds better.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -119,8 +217,11 @@ def starting_long_process(self, message):
+        QApplication.setOverrideCursor(QCursor(Qt.WaitCursor))
+        QApplication.processEvents()
 
+    # FIXME: Should be end_
```
See my comment above.

goanpeca replied:
See my comment above

ccordoba12 replied:
Then this needs to be `after_long_process`

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
+    # FIXME: consider renaming this to setup_plugin, so we can achieve some
+    # language consistency. With this we can then do post_setup_plugin
+    # to refer to things to do after setup, or when visible.
+    # Using register and then post_visible_setup is not clear.
```
I'm -0.5 on this. It'll change a lot of stuff just for some more consistency. At moment  there are no plugin doing something after this step (if I recall things correctly). That's because post_visible_setup is only done in the main window.

goanpeca replied:
>I'm -0.5 on this. It'll change a lot of stuff just for some more consistency. 

Changes won't break anything, they will add the proper deprecation methods for removal on the next major version. I do not plan to keep breaking plugins among versions but we do need to have consistency as that makes developing plugins easier (which is the purpose of this).

>That's because post_visible_setup is only done in the main window.

Some plugins may want to do something after the setup is ready and want to defer it to the post_visible setup method. Like for instance Kite is doing on first install. Right now a signal is used, but we have only that signal which is odd. So it is much better to provide a logic for plugins: 

For example

```python
def register():
    # Create actions and connect to signals, register any plugin types.
    # For example adding a new type to the variable explorer

def setup():
    # Prepare the plugin to start during load

def after_setup():
    # Prepare the plugin to do something after spyder is visble and fully loaded

def unregister():
    # If we wanted to add support for disabling a plugin without a spyder restart this method
    # would perform cleanup and disconnection
```

Now on the main window we can do something like

```
python

def register():
    for plugin in plugins:
        try:
             plugin.register()

def setup():
    for plugin in plugins:
        try:
             plugin.setup()

def after_setup():
    for plugin in plugins:
        try:
             plugin.after_setup()

```

In this process all the hard coded code in main window like defining hard coded plugins and actions ane menus will be moved to the plugin that actually does it or needs it.

This will reduce mainwindow complexity and generalize the process that is currently doing a lot of "manual" work.

goanpeca replied:
@jitseniesen well we don't need. Is just to keep things more consistent.

We could only have setup and post_setup or something... 

But register probably sounds better. I just want to think of things that make the api consistent

jitseniesen replied:
Sorry for being dense, but what is currently inconsistent?

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
-        """
-        Return plain or rich text font used in Spyder.
+    # Path for images relative to the plugin path
+    IMG_PATH = 'images'
```
SpyderPlugin could also require this (for example to use specific images as part of a new toolbar). So please move it to PluginBaseMixin.

goanpeca replied:
I do not really understand this as most plugins should follow a convention.

```
spyder_plugin
    locale/
    images/
    widgets/
    api.py
    plugin.py
    confpage.py
```

The only case I see some utility in this is on packages that are not a spyder_plugin alone but some pakcage that also registers a plugin, and on which case the images folder might not necessarily be inside the plugin folder. 

Having said this yes, I agree that most of the things for the two types of spyder plugins are common except what is strictly necessary for using a dock.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
+        #
+        # Names we use here: register, initialize, on_first_registration
+        # _initialize_plugin_in_mainwindow_layout
+        # Name we use in main window: setup, post_visible_setup
```
* `_initialize` and `_initialize_plugin_in_mainwindow_layout` are hidden, so I see no problem how we name them.
* `register_plugin` and `on_first_registration` have been with us since Pierre's days. We could change their names, but as I said above, that will break a lot of third-party stuff not developed by us for little gain.

goanpeca replied:
>_initialize and _initialize_plugin_in_mainwindow_layout are hidden, so I see no problem how we name them.

Ok

>register_plugin and on_first_registration have been with us since Pierre's days. We could change their names

Doing things the way they always have been done for the sake of it is a really poor justification.

>but as I said above, that will break a lot of third-party stuff not developed by us for little gain

I do not plan to break compatibility between 4 and 5. (as was the case with 2-3, 3-4...)

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
+    # INFO: This is called from mainwindow.py
+    # This name does not convey that this can also (and should) be called
+    # from a non dockable widget. Non dockable widgets should be able to
+    # define a on_first_registration call
```
Why? What could they need to do on first registration?

goanpeca replied:
A kite plugin would want t to show a popup and explain the user something, or open a `kite_tutorial.py`. This is just an example of something that NOW does need this.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
+                return
+            self.set_option('first_time', False)
+
+    # INFO: This is called from mainwindow.py
```
What are these INFO comments?

goanpeca replied:
Things I left while I finish the PR. This is still not done.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -454,6 +744,7 @@ def refresh_plugin(self):
         """
         pass
 
+    # TODO: This method should be deprecated or rethinked
```
This method is very important because it allows to share actions between the plugin and toolbars or main menus. So I don't agree with removing it.

goanpeca replied:
I know that it is being used and I understand how it is used, that is why I say rethink cause it is inconsistently used among different plugins and widgets.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -528,8 +796,159 @@ def toggle_view(self, checked):
         else:
             self.dockwidget.hide()
 
+    # FIXME: What are some examples of this usage?
```
Please look for them. @dalthviz added this method.

goanpeca replied:
@dalthviz could you please point to why we need this? seems like we could most likely define this automatically? I may be wrong so please explain :-)

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -528,8 +796,159 @@ def toggle_view(self, checked):
+    #########################################################################
+    # FIXME: If we turn this into a class attribute, users do not need to
+    # take care of this logic, but only define next to which plugin (by name)
+    # they want to dock their plugin.
```
Ok, agreed.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -528,8 +796,159 @@ def toggle_view(self, checked):
+    # TODO: Rethink this method
+    # A widget should expose its actions in a reliable preditcable way
+    # and the bindning to shortcuts should be performed at the plugin
+    # level only
```
This should be left for another PR.

goanpeca replied:
Yes, agreed ;-)

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -528,8 +796,159 @@ def toggle_view(self, checked):
+
+    # FIXME: This should probably just emit a signal and handle this on
+    # the mainwindow.py directly. Calls to self.main inside a plugin are
+    # Funky!
```
Please leave this method as is or change it in another PR. It took me a hell of lot of time and effort to correctly define the logic for this and check every action and button to verify that it was working correctly.

goanpeca replied:
Not an option anymore.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -476,34 +767,11 @@ def register_plugin(self):
         """
+        # FIXME: We should do this on behalf of the users on the main window
+        # and avoid "silly errors". By doing this we can also move this to
+        # the mixin and unify the information. No nneed to mention docks.
```
I don't know how to do that, but if you know, please do it.

goanpeca replied:
Here the idea is the following:

Widgets should define actions following a defined and more structured method. This can be achieved by having a mixin that all spyder plugins must inherit to provide some facilities to make this.

A widget registers all of its actions (that we want to expose).

Then from the plugin we need to have a method to register shortcuts and actions in a single call and also have a way of registering shortcuts for actions that a widget defines. Since this is currently manual or inoinstent among different widgets, is why I suggest having a mixin to provide some facilities for widgets. 

In this process widgets should never ever use CONF, and any property os behavior that they need to define needs to also follow some standard, either by providing an options dictionary on `__init__`or providing a `setup_XXXX` like some widgets do (code editor to put an example). But again this should follow convention.

jitseniesen replied:
Is `SpyderPluginWidget` a plugin or a widget or both, or do you use plugin and widget interchangeably?

How can widgets react to changes in `CONF` if you provide it as an option dictionary on `__init__`?

goanpeca replied:
>Is SpyderPluginWidget a plugin or a widget or both, or do you use plugin and widget interchangeably?

Well, it is ... both. It is a SpyderPlugin as it defines all the methods of a plugin. When docked it uses the dockwidget attribute but when undocked the plugin itself is the widget of the main window.

>How can widgets react to changes in CONF if you provide it as an option dictionary on __init__?

---

Widgets in general should communicate internal changes to options via signals

`sig_option_changed` so that it can be hook to the plguin (the thing that is aware of the 

And a plugin also should do the same so that thing can bubble down to widgets:

```python
class Plugin
    sig_config_changed   

     def __init__(self):
          self.widgeta = ChildWidgetA(self)
          self.widgetb = ChildWidgetB(self)

          self.sig_config_changed.connect(self.update_options)
          self.widgeta.sig_option_changed.connect(self.update_configuration)
          self.widgetb.sig_option_changed.connect(self.update_configuration)

     def update_options(self, options):
           """This updates options on widgets."""
           for option, value in options.items():
               self.widgeta.update_option(option, value)
               self.widgeta.update_option(option, value)

     def update_configuration(self, options):
           """This updates config with optionsfrom widgets."""
           for option, value in options.items():
               self.config.set(option, value)

class ChildWidgetA:   
    sig_option_changed

class ChildWidgetB:
    sig_option_changed
```

This the general idea that all widgets should follow. To do this I suggest some mixin for ALL spyder widgets so that we provide some basic template (the methods and machinery to update options and inner state, this could simplify some widgets a los as it would remove all the attribute definition) and provide the basic signals.

Thoughts @jitseniesen ?

jitseniesen replied:
Sounds okay. To be honest, I don't remember what I was thinking. I think part of my confusion is that I am not thinking of a widget as the class that should provide an action. For me, a widget is an UI element that should be as dumb as possible. However, Spyder is not written that way (probably because such a pure approach is not practical).

When you say "widgets should never ever use CONF", I guess this does not apply to SpyderPluginWidget even though it is a widget.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -528,8 +796,159 @@ def toggle_view(self, checked):
+    # --- API: available methods ---------------------------------------------
+    # ------------------------------------------------------------------------
+    # FIXME: This SHOULD be private and called from mainwindow.py so users
+    # do not need to know about this on plugin registration
```
Ok.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
+        # of CONF_DEFAULTS for users. And remove the "shortcut"
+        # attribute/special case. It is a little bit more work for the user
+        # but it makes things uniform and simplifies this code. We can add
+        # checks to ensure `switch to {}` matches the right name.
```
This is not work for this PR.

goanpeca replied:
It is if I remove the `self.shortcut` logic plus the shortcut attribute

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -8,69 +8,165 @@
-class BasePlugin(BasePluginMixin):
-    """
-    Basic functionality for Spyder plugins.
+class PluginBaseMixin:
```
PluginBaseMixin sounds odd. It should be BasePluginMixin

goanpeca replied:
Agreed!

jitseniesen commented:
> Sorry to say it but now SpyderPluginWidget is very unreadable (as it was in Spyder 3) by having in the same place a lot of private and public methods and tons of implementation code. That's why I decided to move almost everything to a mixin and just leave an skeleton in `api.py`.

It was not at all clear to me that this was the reason for the rather complicated inheritance structure.

Somebody who wants to write a new plugin needs to know certain information: what attributes and methods to define in the pluign, what the intended behaviour is, and what methods can be called. You are right that this information is hard to find if it is mixed in with implementation details. But I am not sure that your solution is very natural in Python (it works better in a language like C++ that uses header files). I do find it hard to read if I need to go beyond `api/pluign.py` (which at the moment happens quite often).

At first thought, a better way to provide this information is in a document (e.g. a web page). This document can be automatically generated from docstrings in the Python code using a tool such as Sphinx. I do not have experience with Sphinx or other documentation generators in Python, but I expect we can somehow group methods by functionality, indicate what is public or private, or what needs to be defined or overridden in new plugins. People who just need to know the basic information about how to write plugins can then look at the web page, while people who need to look at the code (either Spyder developers or plugin developers who want to do something more advanced) see the code organized in an easy to read fashion.

ccordoba12 commented:
> I do find it hard to read if I need to go beyond api/pluign.py (which at the moment happens quite often).

Ok, thanks for your point of view @jitseniesen. That was the same concern @goanpeca mentioned to me before starting this work. Since there are already two persons against my point of view, I think this refactoring is the way to go.

> At first thought, a better way to provide this information is in a document (e.g. a web page).

Agreed. @goanpeca is also thinking how to do that after finishing this PR.

goanpeca commented:
>@goanpeca, I left my initial review for you.

>Sorry to say it but now SpyderPluginWidget is very unreadable (as it was in Spyder 3) by having in the same place a lot of private and public methods and tons of implementation code. That's why I decided to move almost everything to a mixin and just leave an skeleton in api.py.

It was very unreadable to me and added an unnecessarily complex chain of inheritance that was brittle (as you have experienced with the moving to pyside2)

>very unreadable 

Well you wrote it so there is a bias in there that is difficult to shake.

>So even for us, if we'd like to implement a new core plugin from scratch today, reading this code is really bewildering. So probably your refactoring is more understandable to maintain this same code, but not to develop new plugins.

Writing plugins is not based on reading code, is based on reading documentation. Code should be written to make it easy for developers to continue working with spyder and develop it. In the new system (By the way this PR is not finished and why I left all those FIXME: comments, to gather some feedback before going forward)

I have never written any plugin for any system by going and reading code. I go and find documentation which is going to be part of this fix. Also none of the renaming conventions I am suggesting will be implemented as breaking changes but as deprecation warning to be removed on Spyder 6. Unlike changes from 2 to 3 and 3 to 4 which were completely backwards incompatible.

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -132,61 +233,82 @@ def ending_long_process(self, message=""):
-    # ---------------------------- ATTRIBUTES ---------------------------------
+    # --- API: Optional methods to override ----------------------------------
+    # ------------------------------------------------------------------------
+    def get_plugin_title(self):
```
This should be `get_name_name` as non-dockable (this is a more accurate description of the two types of plugins) also have a localized version. Title is too tied to a concept whereas `get_plugin_name` is more general which serves for both types of plugins

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -132,61 +233,82 @@ def ending_long_process(self, message=""):
-    # One line localized description of the features this plugin implements
-    # Status: Optional
-    DESCRIPTION = None
+    def get_plugin_icon(self):
```
Non-dackable plugins can have a configuration page, (like autopep8) and the need an icon for this. Unless we want to defer this to the confpage class, in the case a plugin can define multiple config pages, which is an idea that we may want to explore as well.

goanpeca commented:
@jitseniesen could you please go over the file and take a look at the `FIXME:` comments? The changes I suggest and the rationale are there. This will imply making some duplicate methods while the FULL deprecation cycle arrives with Spyder 6 but this will allow preserving compatibility and at the same time, provide a new uniform API with more clear naming conventions.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -186,100 +106,164 @@ class SpyderPlugin(BasePlugin):
+    # ------------------------------------------------------------------------
+
+    # FIXME: Also we need to decide if we want to use the word plugin on every
+    # method as a way of consistency. This should be the `get_plugin_option`
```
I'm -1 on this. If these methods belong to a class called Plugin, then it's clear they refer to plugins.

goanpeca replied:
It is not necessary I agree, but some methods have it and some methods do not. I am striving for  consistency which at the end makes stuff predictable for users and developers.

We have `register_plugin`  `get_plugin_name` but then we have methods that do not...

If we add fuzzy matching to spyder (soon to come) doing `SomePlugin.plugin`would provide ALL the methods that the user may use for things so there is a small added benefit here.

We should remove plugin from all things or add it to all but the current state is a messy mixture of different ideas.

jitseniesen replied:
I see. I'd prefer we remove `plugin` from `register_plugin`, `get_plugin_name`.

goanpeca replied:
Removed plugin word

jitseniesen commented:
> @jitseniesen could you please go over the file and take a look at the `FIXME:` comments? The changes I suggest and the rationale are there. This will imply making some duplicate methods while the FULL deprecation cycle arrives with Spyder 6 but this will allow preserving compatibility and at the same time, provide a new uniform API with more clear naming conventions.

Will do, but not now. Probably tomorrow. Until then, some high-level remarks.

- This is a big PR, changing the inheritance chain, renaming some stuff, and other changes. This makes it hard for me to see what is happening.
- Why `PluginBaseMixin`? Is it not possible to have `SpyderPlugin` inherit from only `QObject`, and `SpyderPluginWidget` inherit from `QWidget` and `SpyderPlugin`? Incidentally, I find the name "mix-in" a bit misleading in that `PluginBaseMixin` actually provides most of the functionality.
- I think `SpyderPluginWidget` does a lot by being both a plugin and a widget and I wonder whether composition instead of inheritance would be a better design. That is, have a class `SpyderDockablePlugin` which inherits from `SpyderPlugin` but not `QWidget`, with a member `.dock` of type `SpyderDockWidget` which inherits from `QDockWidget`, thus separating plugin logic from widget logic. That would be quite a major change though.
- I am not convinced about the need for a depreciation cycle because
  - there were more than 3 years between Spyder 3 and Spyder 4;
  - I think our plugin API still needs a fair bit of iterating to get right; 
  - there are also almost no real third-party plugins that I am aware of. 

  If we do have a depreciation cycle, we probably should move to a 6-12 month cycle for major releases and find a way to stick to our plans (e.g. by dropping planned features for the next release instead of postponing the release).

jitseniesen commented on spyder/api/plugins.py:
```diff
@@ -304,115 +288,506 @@ def register_shortcut(self, qaction_or_qshortcut, context, name,
+
+        window.show()
+
+    # FIXME: Now this method sounds like a signal, not a method.
```
I think `on_xxx` sounds like a handler for `xxx`, but I guess `handle_xxx` is more explicit.

goanpeca replied:
Fixed

jitseniesen commented on spyder/api/plugins.py:
```diff
@@ -304,115 +288,506 @@ def register_shortcut(self, qaction_or_qshortcut, context, name,
+        if self._isvisible:
+            self.refresh_plugin()
+
+    # FIXME: This sounds like something that should be public?
```
Agree, I think both the IPythonConsole and the Notebook plugin can use this to (de)activate menu items when appropriate. They also both use `self._options_menu` so perhaps that should be public as well.

goanpeca replied:
Yes

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -304,115 +288,506 @@ def register_shortcut(self, qaction_or_qshortcut, context, name,
+    # str: The message you want to display
+    # int: Amount of time to display the message
+    # FIXME: this is a weird name for a signal
+    sig_show_status_message = Signal(str, int)
```
Yeah, that is what I think is the case.

goanpeca replied:
I have addressed this inconsistency

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -304,115 +288,506 @@ def register_shortcut(self, qaction_or_qshortcut, context, name,
+    # defined once, but every plugin needs to query the config system. This
+    # is a code smell that signals that we should query the property on the
+    # mainwindow and then pass that as an argument to the plugins, or even
+    # better not have this method in here at all!
```
Let me make the fix so it is clearer

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -304,115 +288,506 @@ def register_shortcut(self, qaction_or_qshortcut, context, name,
+            layout.setContentsMargins(*self._default_margins)
+
+    # FIXME: This is called from mainwindow.py and does not need to live
+    # on every plugin.
```
On the main window.

goanpeca commented:
@jitseniesen thanks for the comments, some anwers on inline comments and some here:

>Why PluginBaseMixin? Is it not possible to have SpyderPlugin inherit from only QObject, and SpyderPluginWidget inherit from QWidget and SpyderPlugin? Incidentally, I find the name "mix-in" a bit misleading in that PluginBaseMixin actually provides most of the functionality.

Agree with the naming. The thing is that currently SPyderPluginWidget as you said, is both a spyder plugin and also a widget when undocked. So this decoupling will require some "bigger"changes, but indeed is something good!. `inherit from QWidget and SpyderPlugin`this results in weirdness because of how Qt works. 

>I think SpyderPluginWidget does a lot by being both a plugin and a widget and I wonder whether composition instead of inheritance would be a better design. That is, have a class SpyderDockablePlugin which inherits from SpyderPlugin but not QWidget, with a member .dock of type SpyderDockWidget which inherits from QDockWidget, thus separating plugin logic from widget logic. That would be quite a major change though.

See above, the issue is when a plugin is a separate window and is the central widget of that window. This can be changed. We already have an attribute on the plugin that is the dock and a separate class. We also have the window but it takes the same plugin as the central widget and hence why the plugin in the current model HAS to be a widget as well. This does not need to be the case.

>I am not convinced about the need for a depreciation cycle because

Fair enough, this actually makes it easier for me so 👍

goanpeca commented:
@jitseniesen I can follow your suggestion and have something like:
```python
SpyderPlugin(QObject)
SpyderDockablePlugin(SpyderPlugin)
SpyderDockwidget   -> Already exists
SpyderWidget   -> New widget, JSeb called it PluginCentralWidget

# It would then yield soemthing like

class SpyderDockablePlugin:
    def __init__(self, parent):
        self.dockwidget = SpyderDockwidget()
        self.mainwidget = SpyderWidget()
```

So for example all dockable plugins would need to be split into 2, for example:

```python
PlotsPlugin -> PlotsPlugin(SpyderDockablePlugin)  , PlotsWidget(SpyderWidget)
```

This effectively keeps plugin stuff on the plugin and widget stuff on the widget. Moreover, by enforcing the use of this SpyderWidget we can make the UI be more uniform in terms of toolbars, status widget within a dockable widget, spinners etc. This is also a good opportunity to remove all the custom logic for handling the options button per plugin (a plugin developer should not need to care about that at all. 

This also goes in line with what @jnsebgosselin started on PR https://github.com/spyder-ide/spyder/pull/10626

jitseniesen commented:
I like that idea, but it also sounds like a lot of work ...

goanpeca commented:
>I like that idea, but it also sounds like a lot of work ...

It is, but I am on it already :-)

goanpeca commented:
@jitseniesen  if you could go over the API once again that would be very helpful :-)

@jnsebgosselin I migrated the plots plugin as the first example of using the new API. Could you test it? Also any input on the new API is also welcome :-) 

I have not looked at the tests so far, I may have broken some

I still need to fix the original comments on docstrings from @ccordoba12 and update de code style and fix some of the TODO/FIXME I left

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -304,115 +288,506 @@ def register_shortcut(self, qaction_or_qshortcut, context, name,
+    sig_plugin_registered = Signal()
+    sig_plugin_setup_ready = Signal()
+    sig_plugin_after_setup_ready = Signal()
+    sig_plugin_unregistered = Signal()
```
Removed hem for now

jitseniesen commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1167 @@
+        #     layout.setContentsMargins(*self._default_margins)
+
+    # TODO: Naming, we used refresh, but update_actions might be more clear?
+    # Refresh sounds like painting events...
```
Agree.

goanpeca replied:
Ok

jitseniesen commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1167 @@
+            self.setStyleSheet("QToolBar {border: 0px;}")
+
+
+class SpyderWidget(QMainWindow, SpyderWidgetMixin):
```
Can we have a different name? This name could refer to any widget in Spyder ... I think this was discussed before, but if not, one proposal is `PluginMainWidget`

goanpeca replied:
Yes, makes sense!

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -8,18 +8,35 @@
 main window and call other plugins directly.
+
+There are two types of plugins available:
+* SpyderPluginV2 is a plugin that does not create a new dock/pane on Spyder
```
>Is the plan to rename SpyderPluginV2 to SpyderPlugin at some point? Please document.

Yes, as soon as the migration is over

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+
+    # --- API: Optional attributes ------------------------------------------
+    # ------------------------------------------------------------------------
+    # List of required plugin dependencies
```
````python
    # List of required plugin dependencies classes
    # Example: [PlotsPlugin, IPythonConsolePlugin, ...]
```

jitseniesen replied:
Ah, plugin dependencies mean dependencies that are a plugin, not dependencies of this plugin.

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+
+    # Use this signal to inform another plugin that configuration changed.
+    sig_option_changed = Signal(str, object)
+    sig_toggle_view_changed = Signal(bool)
```
This should only go in the DockablePlugin. It refers to the toggle view action of a QDockWidget, in this case it refers to the toggle view action of the SpyderDockWidget.

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+    CONF_WIDGET_CLASS = None
+
+    # Path for images relative to the plugin path
+    IMG_PATH = 'images'
```
A python package might be a python package and also include a spyder plugin  (or several) in this case the package may be using images from a global folder so in order for the plugin to also have access to these images a relative path might be used.

I am not sure how this has been used, @ccordoba12 might have more information

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+        """
+        Return the plugin system path.
+        """
+        return self._plugin_path
```
We could, but since PyQt does not follow this pattern a lot, it feels more natural to provide similar getters on our API. Is more of a consistency issue.

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+    def create_action(self, name, text, icon=None, tip=None, toggled=None,
+                      triggered=None, data=None, menu_role=None,
+                      context=Qt.WindowShortcut, shortcut_context=None,
+                      expose=True, disable_if_not_visible=True, parent=None):
```
>This function does not actually set a shortcut, right? How is a plugin supposed to do that?

There is no need to set shortcuts right now. We only create actions with this (and similar methods) and these are then exposed automagically as possible shortcuts. If a shortcut is found in the default config then it is assigned, otherwise is left blank for the user to use.

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+        QIcon
+            QIcon instance
+        """
+        raise NotImplementedError('A plugin description must be defined!')
```
Thanks!

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+    def check_compatibility():
+        """
+        This method can be reimplemented to check compatibility of a
+        plugin for a given condition.
```
Fixed

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+        Or to perform actions that are needed only the first time this is
+        installed.
+
+        This method is calle after the main window is visible
```
Thanks

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+            with PyQt4'). It will be shown at startup in a QMessageBox.
+        """
+        valid = True
+        message = _('')
```
Thanks

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+    OPTIONAL = None
+
+    # This is the a widget (or any non widget class) to store behavior and
+    # actions. Subclass SpyderWigdetMixin
```
I need to elaborate this a bit more you are right. I will try to add some text in there.

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+
+    # --- API: Available signals ---------------------------------------------
+    # ------------------------------------------------------------------------
+    sig_toggle_view_changed = Signal(bool)
```
My bad, should only be here

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+    # {plugin.NAME: plugin_instance, ...}
+    _PLUGINS = None
+
+    def __init__(self, parent, configuration=None):
```
>You don't want __init__ to be overridden in subclasses? That seems quite unnatural; is that necessary?

Actually no. The idea is not to override the init, only here because we need to set some  initialization for dockable plugins on top of the "normal" plugin initialization.

The PlotsPlugin looks like:

```python
class PlotsPlugin(SpyderDockablePlugin):
    """
    Plots plugin.
    """

    NAME = 'plots'
    REQUIRES = [IPythonConsole]
    WIDGET_CLASS = PlotsWidget
    TABIFY = VariableExplorer
    CONF_SECTION = NAME  # FIXME: Redundant? move to use NAME for this as well
    CONF_FILE = False
    DISABLE_ACTIONS_WHEN_HIDDEN = False

    def get_name(self):
        return _('Plots')

    def get_description(self):
        return _('Display, explore and save console generated plots.')

    def get_icon(self):
        return self.create_icon('hist')

    def register(self):
        # Plugins
        ipyconsole = self.get_plugin(IPythonConsole)

        # Signals
        widget = self.get_widget()
        ipyconsole.sig_shellwidget_id_changed.connect(
            widget.set_shellwidget_from_id)
        ipyconsole.sig_shellwidget_process_started.connect(
            widget.add_shellwidget)
        ipyconsole.sig_shellwidget_id_process_finished.connect(
            widget.remove_shellwidget_from_id)

    # Exposed API
    def current_widget(self):
        return self.get_widget().current_widget()
```

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+    """
+    # --- API: Mandatory attributes -------------------------------------------
+    # ------------------------------------------------------------------------
+    # This is the main widget of the dockable plugin. Subclass SpyderWigdet
```
Fixed!

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1167 @@
+        )
+        # We use this instead of the QDockWidget.toggleViewAction
+        toggle_action = self.create_action(
+            name='switch to plots',  # TODO: Move to use names with `_`
```
Typo!

goanpeca replied:
Yep!

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1167 @@
+            icon=ima.icon('tooloptions'),
+        )
+        self.add_corner_widget('options_button', self._options_button)
+        self.add_corner_widget('spinner', self._spinner)
```
It is something that was added to the variable explorer recently on this PR

https://github.com/spyder-ide/spyder/pull/11814

<img width="653" alt="Screen Shot 2020-03-31 at 10 53 30" src="https://user-images.githubusercontent.com/3627835/78047383-e4614280-733d-11ea-8f0e-89c1ca31384e.png">

Now we added that as default for all widgets. 

The Find in files plugin also has a spinner, but is located on another place, we will make things uniform now with this.

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1167 @@
+
+    def get_sections(self):
+        """
+        Return name for menu as unique identifier.
```
Fixed!

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1167 @@
+        self._dirty = True
+        self._ordered_actions = []
+
+    def get_name(self):
```
I left both to decide what makes more sense. So I could use some input here @ccordoba12

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+
+    def get_actions(self):  # Old name: get_plugin_actions
+        """
+        Return a dictionary of actions exposed by the plugin and child widgets.
```
Exposed actions are actions created by the self.create_action method,
        within this plugin or on widgets that use the SpyderWidgetMixin.

        Any time self.create_action is used, the widget stores the actions on
        a dictionary {action_name: action,...}.

        This method by default returns the actions created on the
        PluginMainWidget only, but that method can be override to collect all
        actions in child widgets.

goanpeca commented on spyder/plugins/plots/plugin.py:
```diff
@@ -155,3 +186,415 @@ def apply_plugin_settings(self, options):
+# New API!
+#
+##############################################################################
+class PlotsWidget(SpyderWidget):
```
@jitseniesen please look at how this `PluginMainWidget` looks like.

goanpeca commented on spyder/plugins/plots/plugin.py:
```diff
@@ -155,3 +186,415 @@ def apply_plugin_settings(self, options):
+            widget.zoom_out()
+
+
+class PlotsPlugin(SpyderDockablePlugin):
```
@jitseniesen please look at how this `PlotsPlugin` looks like. it is using the new API.

goanpeca commented on spyder/plugins/plots/plugin.py:
```diff
@@ -155,3 +186,415 @@ def apply_plugin_settings(self, options):
+
+    def setup(self, options):
+        # Menu actions
+        mute_action = self.create_action(
```
All these actions here @jitseniesen are exposed as shortcuts without having to register a shortcut manually.

goanpeca commented:
>Looks good, most of my comments are about where things are not clear to me or the documentation can be improved. I look forward to trying it out and seeing how well it works in practice.

Thanks for the review @jitseniesen , addressed your comments and made the updates. Pinged you in a couple of lines so you can see how the new API applies to the plots widget

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -250,6 +250,157 @@ class MainWindow(QMainWindow):
+        messageBox.show()
+
+    def register_plugin(self, plugin):
+        """
```
This is where the new plugins are registered on the mainwindow.

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -250,6 +250,157 @@ class MainWindow(QMainWindow):
+            context = (getattr(action, 'shortcut_context', plugin.NAME)
+                       or plugin.NAME)
+            self.register_shortcut(action, context, action_name)
+
```
Here is where "exposed actions" are registered as shortcuts.

I need to do some cleanup but the idea is the following.

- Actions should be created once (not destroyed and recycled on popup of menus etc,...)
- To create an action the user must user `self.create_action` by using the SpyderWidgetMixin.
- A PluginMainWidget will collect any actions defined in subwidgets (if defined) and expose them in the `get_actions` method. This could be even automated so that we find all children widgets and check if they define these actions. By doing it like this, users do not need to worry about exposing these actions manually but the API does it for them
- Any action created this way is now exposed as a possible shortcut automatically without manual registration. If an option is found in the config then it is assigned otherwise is left with an empty shortcut. By doing it like this the `name` of actions is now also used for the `name`of the shortcut and by doing that we already have the translation of the shortcut via the action, plus we now never need to worry about these shortcuts.

Is it more clear now?

jitseniesen replied:
Yes, thank you. If you can write part of this in the doc string of `create_action` and `get_actions`, that would be grand.

goanpeca replied:
Will do!

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+    # Example: ['plots', 'ipython_console', ...]
+    REQUIRES = None
+
+    # List of required plugin dependencies names
```
required -> optional?

And why does it mean that a plugin has optional dependencies? Could give update the comment above with an example?

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+
+    # This is the a widget (or any non widget class) to store behavior and
+    # actions. Subclass SpyderWigdetMixin
+    WIDGET_CLASS = None
```
I don't get this. Why could this be needed for SpyderPlugin's? Besides, the comment is confusing because I don't understand what you mean by `Subclass SpyderWigdetMixin`.

goanpeca replied:
The Working Directory plugin creates a New widget, a Toolbar. In those cases it makes sense to put that widget here. If a Plugin is creating "several" non dock widgets, a status bar widgets, a menu, a new toolbar, those widgets could (or should be grouped) within some "parent"widget.

jitseniesen replied:
It would help if the comment describes what is done with the class stored in `WIDGET_CLASS`.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+    # config system (i.e. in spyder.ini)
+    CONF_SECTION = None
+
+    # Use separate configuration file for plugin
```
```suggestion
    # Use a separate configuration file for the plugin
```

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+            self._PLUGINS = {}
+
+        if self.NAME is None:
+            raise Exception('A Spyder Plugin must define a `NAME`!')
```
Why are you using such a broad exception here? Perhaps `RuntimeError` would be better?

goanpeca replied:
Agreed, all the exceptions are currently broad and need to be updated.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+            raise Exception('A Spyder Plugin must define a `NAME`!')
+
+        if self.NAME in self._PLUGINS:
+            raise Exception('A Spyder Plugin with NAME="{}" already exists!'
```
The same here.

goanpeca replied:
Agreed, all the exceptions are currently broad and need to be updated.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+
+    def _unregister(self):
+        """
+        Disconnect signals and clean up the plugin to be able to stop a plugin
```
```suggestion
        Disconnect signals and clean up the plugin to be able to stop it
```

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+        Disconnect signals and clean up the plugin to be able to stop a plugin
+        while Spyder is running.
+        """
+        self.sig_option_changed.disconnect()
```
This probably needs to be inside a try/except TypeError. This full `disconnect` frequently generates that error.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+
+    # --- API: available methods ---------------------------------------------
+    # ------------------------------------------------------------------------
+    # There is no need to set shortcuts right now. We only create actions with
+    # this (and similar methods) and these are then exposed as possible
+    # shortcuts on plugin registration in the main window.
+    # If a shortcut is found in the default config then it is assigned,
+    # otherwise is left blank for the user to use.
```
I think this big comment should be part of the method docstring instead because it'd be useful to generate docs for this API later on).

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+
+    def get_path(self):
+        """
+        Return the plugin system path.
```
```suggestion
        Return the plugin's system path.
```

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+            unique identifiable name for the action
+        text: str
+           Localized text for the action
+        ...
```
Please change this `...` for a `TODO`, so you don't forget to address it at the end.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+
+    def get_plugin(self, plugin_name):
+        """
+        Return a plugin instance by providing the plugin name.
```
```suggestion
        Return a plugin instance by providing the plugin's name.
```

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+
+    # Store reference to registered plugins in a dictionary
+    # {plugin.NAME: plugin_instance, ...}
+    _PLUGINS = None
```
Shouldn't this be an attribute of the main window? It seems you're using it to access the plugins registered before, but I don't understand how you're doing that if this an attribute of the class.

goanpeca replied:
Yes at some point it will need to be a part of the `PluginManager` which will live in the MainWindow. But that will only be posible after ALL plugins are migrated

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+        Returns
+        -------
+        Dictionary of {str: object}
+
```
```suggestion
```

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+            section = self.CONF_SECTION if section is None else section
+            self._conf.set(section, str(option), value)
+
+    # TODO: Configuration vs Options vs settings?
+    # We could use options for widgets and decide on configuration vs
+    # settings
```
You could change this method's name to `apply_conf` or `apply_config`, to be more consistent.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+
+    def update_font(self):
+        """
+        This must be reimplemented by plugins that need to adjust their fonts.
```
Please give here an example of the core plugins that uses this.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+        """
+        pass
+
+    def update_style(self):
```
What this refers to? Please clarify it in the docstrings below, along with an example.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+        """
+        pass
+
+    def update_translation(self):
```
This is even more vague for me. Please clarify it too.

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+    # ------------------------------------------------------------------------
+    # Define next to which widget we want to tabify this plugin
+    # Example: 'plots'
+    # TODO: Make a list of plugin names? Or global dock sections
```
A list is better because we don't plan to add much more new plugins later on.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+        widget.update_actions()
+
+        # TODO: Streamline this
+        # Render all toolbars as a final separate step on the main window
+        # in case some plugins want to extend a toolbar. Since the rendering
+        # can only be done once!
```
I agree with this. It should be done in post_visible_setup

ccordoba12 commented on spyder/api/plugins.py:
```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+                                name,
+                                list(app_toolbars.keys())))
+
+    def add_item_to_application_menu(self, item, menu, section=None,
```
This should be possible with an SpdyerPlugin, right? I mean, a plugin could add an entry to a toolbar or a menu without having a graphical counterpart (as I think we discussed before).

goanpeca replied:
Agreed, will move!

ccordoba12 commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1478 @@
+        return tuple(self._sections)
+
+
+class MainWidgetMenu(SpyderMenu):
```
Does this contain all the hacks we've added during the years to better support macOS? I don't see them in code below, so that would a terrible regression.

goanpeca replied:
Not sure what you mean. I saw some hacks on the `create_action` qthelpers which are still being used since I call that method.

For  add_actions I saw no hack.

ccordoba12 commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1478 @@
+            self._dirty = False
+
+
+class ApplicationMenu(SpyderMenu):
```
I don't understand the purpose of this empty class.

goanpeca replied:
Ensure that the methods will receive the appropriate Menu. This menu will need to perform some extra checks. I will add this as soon as I start working with a plugin that creates a lot of actions inside an app menu (like editor or console)

ccordoba12 commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1478 @@
+    """
+
+
+class ApplicationToolBar(QToolBar):
```
Nor this one.

goanpeca replied:
Same here, I needed the placeholders but this functionality will come in a few days.

ccordoba12 commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1478 @@
+
+# --- Spyder Mixins
+# ----------------------------------------------------------------------------
+class SpyderOptionMixin:
```
This is missing a docstring.

ccordoba12 commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1478 @@
+        self._options = options
+
+
+class SpyderToolBarMixin:
```
Missing docstring too.

ccordoba12 commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1478 @@
+        return self._toolbuttons
+
+
+class SpyderMenuMixin:
```
Missing docstring.

ccordoba12 commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1478 @@
+        return self._menus
+
+
+class SpyderActionMixin:
```
Missing docstring.

ccordoba12 commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1478 @@
+        self._options_button.setMenu(options_menu)
+        options_menu.aboutToShow.connect(self._update_actions)
+
+        # Show icons in Mac plugin menus
```
Show -> Hide ?

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1478 @@
+        self._menus[name] = menu
+        return menu
+
+    def _add_auxiliary_toolbar(self, toolbar, location):
```
When is this necessary?

goanpeca replied:
Ideally a panel should only have a single set of tool buttons or widgets on top, like the variable explorer, the plots pane etc.

Some plugins are a bit "weird" and verbose and add "two lines" (like the pylint one) or 3 lines "find in files" 

These are now all ToolBars, so anythong beyondthe main one, is an auxiliary one.

goanpeca replied:
On this subject, part of reviewing the UI once this migration occurs is to take some decisions on what we want and enforce that in an uniform way for all plugins.

Some things to take into account:
- [ ] Names vs no names next to toolbuttons (I would say, no names for all!)
- [ ] Simplification of plugins with too many buttons, options or "lines of widgets"on the top of the pane (like find in files...)
- [ ] Remove duplicate buttons like Start/Stop, and replace by a single one (like find in files now does)

ccordoba12 commented on spyder/app/mainwindow.py:
```diff
@@ -250,6 +250,204 @@ class MainWindow(QMainWindow):
+                # Do we really need this?
+                # plugin.on_first_registration()
+            except NotImplementedError:
+                return
```
Should this be `return` or `pass`?

goanpeca replied:
Pass!

ccordoba12 commented on spyder/app/mainwindow.py:
```diff
@@ -250,6 +250,204 @@ class MainWindow(QMainWindow):
+            conf_widget.initialize()
+            return conf_widget
+
+
```
```suggestion

```

ccordoba12 commented on spyder/app/mainwindow.py:
```diff
@@ -3203,35 +3505,66 @@ def set_prefs_size(self, size):
-            keyseq = QKeySequence(CONF.get_shortcut(context, name,
-                                                    plugin_name))
             try:
-                if isinstance(qobject, QAction):
-                    if sys.platform == 'darwin' and \
-                      qobject._shown_shortcut == 'missing':
-                        qobject._shown_shortcut = keyseq
-                    else:
```
Why are you removing this logic? This took us years to get it right.

goanpeca replied:
It is moved bellow.

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -370,6 +391,9 @@ class SpyderPluginWidget(SpyderPlugin, BasePluginWidget):
 
     # Path for images relative to the plugin path
     # Status: Optional
+    # A python package might be a python package and also include a (or
```
Fixed

goanpeca commented:
>This looks much better than before, but still needs work.

Sure

>I left a bunch of suggestions (to improve docstrings and comments) and several comments (specially in api/widgets.py).

Thanks!

>I still don't understand the purpose of WIDGET_CLASS in SpyderPluginV2 from the comments and docstrings. I mean, it makes sense in SpyderDockablePlugin, but it seems you're assuming that SpyderPluginV2 should always declare a WIDGET_CLASS, which doesn't make sense to me.

Yes, I can understand why this is a source of confusion, I can show you some examples tomorrow on how it is used by other plugins.

?I don't understand the difference between PluginWidget and PluginMainWidget. Why do we need the first?

This is related to the previous question. Plugins that are not docks but want to provide some functionality need to define some class that contains the functionality, the definitions of actions and overall a Class/Object/Widget to group functionality.

>The switch_to_plugin functionality seems to be broken in this PR (you left several TODO's about it and unconnected signals).

I did not connect some signals and left them commented to be sure I did not break anything, but AFAICT, it is working as expected. without those commented lines.

>Please don't rebase nor squash this PR until we decide it's time to merge it, so I can see how you address my comments in the following commits.

Ok

>Please stop porting our plugins to the new API until this is finally merged.

Ok

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -256,6 +257,237 @@ class MainWindow(QMainWindow):
+        """
+        Return a plugin instance by providing the plugin class.
+        """
+        # TODO: Handle by class or by name
```
`TODO`are future work no for this PR, `FIXME`are things that need to be fixed for the PR

goanpeca replied:
Fixed

goanpeca commented on spyder/api/mixins.py:
```diff
@@ -0,0 +1,669 @@
+
+                actions.append(None)
+
+            # FIXME: Remove any consecutive Nones
```
No need it seems when no items are between separators, Qt takes care of that for us

goanpeca commented on spyder/api/mixins.py:
```diff
@@ -0,0 +1,669 @@
+
+class SpyderOptionMixin:
+    """
+    This mixin provides option handling tools for any widget that needs to
+    track options, their state and methods to call on option change.
```
Not necessarily, some options might be tied to the configuration, some other might be specific to the widget, instead of using a CLASS_ATTRIBUTE.

goanpeca replied:
Fixed

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+            ApplicationMenus.Help: self._main.help_menu_actions,
+        }
+        actions = app_menu_actions[menu.name]
+        actions.append(None)
```
Once the plugins are migrated, there wont be any need for doing this, since that will be handled at the SpyderMenu level, so probably not worth it as this is temporal code.

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        self._layout = QHBoxLayout()
+        self.setLayout(self._layout)
+
+        # FIXME: Decide on a number
```
I don't know, but it defines how close together the widgets in the corner will be, this value is good enough for now

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        self._icon_size = QSize(16, 16)
+
+        # Setup
+        self.setObjectName("main_widget_toolbar_{}".format(
+            str(uuid.uuid4())[:8]))
```
I don't know, it was there from the PR that @jnsebgosselin did.

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+
+    Notes
+    -----
+    All Spyder non dockable plugins define a plugin widget that must subclass
+    this.
```
Because Plugins are no longer widgets so they cannot mix implementations of plugins and widgets like before (which was quite a mess)

Even if a Plugin does not have a "Physical visible entity" it still needs to provide a Class to store the functionality and to be able to have a Class that is testable and usable without needing to use a Plugin.

A completion Plugin will live on this "Class" it HAS to be a QObject to be able to have signals, but if this holder Class is not also a QWidget it cannot parent child widgets like a status bar widget, so for instance in the case of completions.

An example:

```python
# spyder.plugins.kite.widgets.main_widget.py
class KiteCompletionClientWidget(PluginWidget, SpyderCompletionMixin):
      DEFAULT_OPTIONS = {}

      def __init__(self, ...)
            self.kite_status = KiteStatusWidget(self)

    # SpyderWidgetMixin API
    def setup(self, options=DEFAULT_OPTIONS):
            # Create actions, menus, etc

     def on_option_update(self, options=DEFAULT_OPTIONS):
            ...

     def update_actions(self, options=DEFAULT_OPTIONS):
            ...

      # The rest of the Completion API request, broadcast start etc....
     ....


# spyder.plugins.kite.plugin.py
class Kite(SpyderPluginV2):
     WIDGET_CLASS = KiteCompletionClientWidget
```

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+
+    def show_status_message(self, message, timeout):
+        """
+        Show a status message in the Spyder widget.
```
They could, if showing information on the Application status bar would be too cumbersome.

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        window.setWindowIcon(self.get_icon())
+        window.setCentralWidget(self)
+        window.setWindowTitle(self.get_title())
+        window.resize(self.size())
```
Using `window.setAttribute(Qt.WA_DeleteOnClose)` usually leads to typerrors when C has destroyed the thing but it still lives in Python.

goanpeca replied:
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        # self.sig_refresh_requested.emit()
+
+    @Slot()
+    def close_window(self):
```
I don't understand your comment, please clarify.

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+
+        self.is_visible = enable
+
+        # TODO: Handle this option
+        # if getattr(self, 'DISABLE_ACTIONS_WHEN_HIDDEN', None):
+        #     toggle_actions(self.get_actions(), visible)
+        #     toggle_actions(self._plugin_actions, visible)
+
+        # self.is_visible = enable and visible
```
>Please fix this. As you can see in the last line here, enable and visible were two different things in the old API. I don't understand why you're changing that. Also, that could have subtle but serious implications for several panes.

The logic in the old API was not great, otherwise this major refactoring would not have been needed. I have not need to use this code in the testing I made.

So please test any of the over 10 PRs available with different plugins to ensure things work,.

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+
+        # Setup
+        dock.setObjectName(self.__class__.__name__ + '_dw')
+        dock.setWidget(self)
```
They live on the Dock

dock.setAllowedAreas(dock.ALLOWED_AREAS)
dock.setFeatures(dock.FEATURES)

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        dock.topLevelChanged.connect(self._on_top_level_change)
+
+        # TODO: Change name of signal and connect? when is this used?
+        # dock.sig_plugin_closed.connect(self.close_dock)
```
Most of the things I left commented out are no longer needed. I will erase them

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -256,6 +257,237 @@ class MainWindow(QMainWindow):
+                if isinstance(plugin, SpyderDockablePlugin) and plugin.TABIFY:
+                    next_to_plugin = self.get_plugin(plugin.TABIFY)
+                    self.tabify_plugins(plugin, next_to_plugin)
+            except NotImplementedError:
+                pass
```
This is a little more complex at this point and should not hold this PR back. Since the new Spyder API will allow for users to start spyder with only the plugins they want, using a "Plugin" on Tabify will not cut it. Some options:

1.) We make that Tabify Plugin a REQUIRES dependency which feels very odd.
2.) We tabify on "abstract locations" that are defined for the standard layouts
3.) We do not make TABIFY a required plugin, and if not found, we use a default location that will always exist, for example, the Internal Console.

goanpeca replied:
For now using option 3

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+    # This variable helps translate CONF to options when the option comes
+    # from another plugin.
+    # Example:
+    # CONF_FROM_OPTIONS = {'widget_option': ('section', 'option'), ...}
```
Done

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+    # --- API: Signals -------------------------------------------------------
+    # ------------------------------------------------------------------------
+
+    # Use this signal to inform another plugin that configuration changed.
```
Fixed

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+        section: str
+            TODO
+
+        Notes
+        -----
+        * Use sig_option_changed to call this method from widgets of the
+          same or another plugin.
+        * CONF_SECTION needs to be defined for this to work.
```
Fixed

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+            Message to show in the status bar when the long
+            process starts.
+        """
+        if self.get_widget() is not None:
+            self.get_widget().start_spinner()
```
Fixed

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+        self.show_status_message(message, timeout=2000)
+        QApplication.processEvents()
+
+        if self.get_widget() is not None:
+            self.get_widget().stop_spinner()
```
Fixed

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+        -----
+        1. Actions should be created once. Creating new actions on menu popup
+           is *highly* discouraged.
+        2. To create an action the user must user this method on
+           SpyderWidgetMixin.
```
Fixed

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+        pass
+
+    # FIXME: Better name
+    def on_visible(self):
```
Fixed

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+        """
+        This must be reimplemented by plugins that need to adjust their fonts.
+
+        FIXME: Give examples of core plugins that use this.
```
Fixed###

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -533,3 +561,951 @@ def set_ancestor(self, ancestor):
+    # this signal. By checking/unchecking the option for a pane inside
+    # the view menu in the application this is triggered.
+    sig_toggle_view_changed = Signal(bool)
+    sig_switch_to_plugin_requested = Signal(object, bool)
```
Fixed

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -256,6 +257,237 @@ class MainWindow(QMainWindow):
+
+        if not is_compatible:
+            self.show_compatibility_message(message)
+            return None
```
Fixed

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -256,6 +257,237 @@ class MainWindow(QMainWindow):
+
+        if isinstance(plugin, SpyderDockablePlugin):
+            # Add dockwidget
+            logger.info("Adding dockwidget for {}...".format(plugin.NAME))
```
Fixed

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -256,6 +257,237 @@ class MainWindow(QMainWindow):
+        for signal in signals:
+            try:
+                signal.disconnect()
+            except Exception:
```
Fixed

goanpeca commented on spyder/api/mixins.py:
```diff
@@ -0,0 +1,669 @@
+            self._toolbars = OrderedDict()
+
+        if name in self._toolbars:
+            raise Exception('!')
```
Fixed

goanpeca commented on spyder/api/mixins.py:
```diff
@@ -0,0 +1,669 @@
+            self._toolbars = OrderedDict()
+
+        if name not in self._toolbars:
+            raise Exception('!')
```
Fixed

goanpeca commented on spyder/api/mixins.py:
```diff
@@ -0,0 +1,669 @@
+
+    # TODO: Support adding items to menu name (enum?)
+    # Add default actions to options menu
+    def add_item_to_menu(self, action_or_menu, menu, section=None,
+                         before=None):
```
Fixed

goanpeca commented on spyder/api/mixins.py:
```diff
@@ -0,0 +1,669 @@
+        -----
+        1. Actions should be created once. Creating new actions on menu popup
+           is *highly* discouraged.
+        2. To create an action the user must user this method on this Mixin.
```
Fixed[](url)

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        This method is called when the an option is set with the `set_option`
+        or `set_options` method from the OptionMixin.
+        """
+        raise NotImplementedError
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+
+        Exposed actions are actions created by the self.create_action method.
+        """
+        raise NotImplementedError
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+
+        if self.dockwidget is not None:
+            # TODO: need a signal!
+            # self.set_ancestor(self.main)
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+            self.dockwidget.setVisible(True)
+            self.dockwidget.raise_()
+            self._update_actions()
+            # self.sig_switch_to_plugin_requested.emit()
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        window.show()
+
+        # Signals
+        # TODO:
+        # self.sig_refresh_requested.emit()
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+            interface.
+        """
+        if name in self._toolbars:
+            raise Exception('!')
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+            The main toolbar of the plugin that contains the options button.
+        """
+        if name not in self._toolbars:
+            raise Exception('TODO:')
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        1. Actions should be created once. Creating new actions on menu popup
+           is *highly* discouraged.
+        2. To create an action the user must user this method on this Mixin.
+        3. The PluginMainWidget will collect any actions defined in subwidgets
+           (if defined) and expose them in the get_actions method at the plugin
+           level.
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        -----
+        1. Actions should be created once. Creating new actions on menu popup
+           is *highly* discouraged.
+        2. To create an action the user must user this method on this Mixin.
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+
+        # TODO: Should provide the right section?
+        bottom_section = OptionsMenuSections.Bottom
+        self.add_item_to_menu(
+            self.undock_action,
+            self._options_menu,
+            section=bottom_section,
+        )
+        self.add_item_to_menu(
+            self.close_action,
+            self._options_menu,
+            section=bottom_section,
+        )
+        self.add_item_to_menu(
+            self.dock_action,
+            self._options_menu,
+            section=bottom_section,
+        )
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+            self._spinner,
+        )
+
+        # TODO: Check if the central widget added is a QTabWidget or similar
+        # and hide the main toolbar, in which case only a corner widget is
+        # used
```
Yes, fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+    DEFAULT_OPTIONS = {}
+
+    # Signals
+    sig_option_changed = Signal(str, object)
+    sig_redirect_stdio_requested = Signal(bool)
+    sig_refresh_requested = Signal()  # FIXME:
+    sig_toggle_view_changed = Signal(bool)
+    sig_update_ancestor_requested = Signal()  # FIXME:
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+    # ------------------------------------------------------------------------
+    def get_title(self):
+        """
+        Return the title that will be displayed on dockwidget or window title.
```
Correct

goanpeca replied:
Oops, removing!

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+    DEFAULT_OPTIONS = {}
+
+    # Signals
+    sig_option_changed = Signal(str, object)
+    sig_redirect_stdio_requested = Signal(bool)
+    sig_refresh_requested = Signal()  # FIXME:
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+
+
+# --- Constants
+# ----------------------------------------------------------------------------
```
Ok

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+
+# --- Spyder Widgets
+# ----------------------------------------------------------------------------
+class SpyderWindowWidget(QMainWindow):
```
Ok

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+    def _render(self):
+        """
+        Create the menu prior to showing it. This takes into account sections
+        and location of menus. It also hides consecutive separators if found.
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+            return self._widgets[name]
+
+
+class MainWidgetMenu(SpyderMenu):
```
Fixed

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+            self._dirty = False
+
+
+class ApplicationMenu(SpyderMenu):
```
>If this going to be used to at the app level and not at panes level

I do not understand. An application menu will be created with the Plugin API

so a SpyderPluginV2 or a SpyderDockablePlugin will do something like

`self.create_application_menu('name', _('Localized title'))`  

Devs are not expected to use this directly.

goanpeca replied:
Ok, need to think about this

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+    """
+
+
+class ApplicationToolBar(QToolBar):
```
Ok, need to think about this

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+    """
+
+
+class MainWidgetToolbar(QToolBar):
```
Done

goanpeca commented on spyder/api/widgets.py:
```diff
@@ -0,0 +1,1176 @@
+        Take into account the existence of a corner widget when adding a new
+        action in this toolbar.
+        """
+        pass
```
Not really it is more an override to prevent its direct usage.

ccordoba12 commented on spyder/api/widgets/__init__.py:
```diff
@@ -372,22 +103,36 @@ def __init__(self, name, plugin, parent=None, options=DEFAULT_OPTIONS):
+    def on_option_update(self, option, value):
         """
-        raise NotImplementedError('PluginWidget must define `setup`!')
+        This method is called when the an option is set with the `set_option`
```
```suggestion
        This method is called when an option is set with the `set_option`
```

ccordoba12 commented on spyder/api/widgets/__init__.py:
```diff
@@ -401,17 +146,28 @@ class PluginMainWidget(QMainWindow, SpyderWidgetMixin, SpyderToolBarMixin):
+
+    # The action that toggles the visibility of a dockable plugin fires
+    # this signal. This is triggered by checking/unchecking the option for
+    # a pane in the View menu.
```
```suggestion
    # a pane in the `View > Panes` menu.
```

ccordoba12 commented on spyder/api/widgets/__init__.py:
```diff
@@ -981,10 +722,7 @@ def create_window(self):
-        # TODO:
-        # self.sig_refresh_requested.emit()
-
+    # FIXME: Check what changed in this method
```
Please address this before merging this PR.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/widgets/__init__.py:
```diff
@@ -1043,12 +777,10 @@ def change_visibility(self, enable, force_focus=None):
         self.is_visible = enable
 
-        # TODO: Handle this option
+        # TODO: Pedning on plugin migration that uses this
```
```suggestion
        # TODO: Pending on plugin migration that uses this
```

ccordoba12 commented on spyder/api/widgets/__init__.py:
```diff
@@ -1131,29 +860,33 @@ def setup(self, options):
         """
-        raise NotImplementedError
+        raise NotImplementedError(
+            'A PluginMainWidget subclass must define a `update_actions` '
```
```suggestion
            'A PluginMainWidget subclass must define an `update_actions` '
```

ccordoba12 commented on spyder/api/widgets/__init__.py:
```diff
@@ -1131,29 +860,33 @@ def setup(self, options):
         """
-        raise NotImplementedError
+        raise NotImplementedError(
+            'A PluginMainWidget subclass must define a `on_option_update` '
```
```suggestion
            'A PluginMainWidget subclass must define an `on_option_update` '
```

ccordoba12 commented on spyder/api/widgets/menus.py:
```diff
@@ -0,0 +1,151 @@
+
+class ApplicationMenu(SpyderMenu):
+    """
+    Spyder Main Window application Menus.
```
```suggestion
    Spyder Main Window application Menu.
```

ccordoba12 commented on spyder/api/widgets/mixins.py:
```diff
@@ -608,10 +543,12 @@ def get_actions(self):
-           (if defined) and expose them in the get_actions method at the plugin
-           level.
+        2. Actions can be created directly on a PluginMainWidget or
+           PluginWidget subclass. Child widgets can also create actions, but
+           they need to subclass SpyderWidgetMixin.
```
```suggestion
           PluginMainContainer subclass. Child widgets can also create
           actions, but they need to subclass SpyderWidgetMixin.
```

ccordoba12 commented on spyder/api/widgets/mixins.py:
```diff
@@ -608,10 +543,12 @@ def get_actions(self):
+        2. Actions can be created directly on a PluginMainWidget or
+           PluginWidget subclass. Child widgets can also create actions, but
+           they need to subclass SpyderWidgetMixin.
+        3. The PluginMainWidget or PluginWidget will collect any actions
```
```suggestion
        3. PluginMainWidget or PluginMainContainer will collect any actions
```

ccordoba12 commented on spyder/api/widgets/toolbars.py:
```diff
@@ -0,0 +1,249 @@
+        self._section_items = OrderedDict()
+        self._title = title
+
+    def add_item(self, action_or_widget, section=None, before=None):
```
This method is missing a docstring.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/widgets/toolbars.py:
```diff
@@ -0,0 +1,249 @@
+
+    def _render(self):
+        """
+        Create the toolbar taking into account the sections and locations.
```
```suggestion
        Create the toolbar taking into account sections and locations.
```

ccordoba12 commented on spyder/api/widgets/toolbars.py:
```diff
@@ -0,0 +1,249 @@
+    """
+    Spyder Widget toolbar class.
+
+    A toolbar used in Spyder Widgets to add internal toolbars
```
```suggestion
    A toolbar used in Spyder dockable plugins to add internal toolbars
```

ccordoba12 commented on spyder/api/widgets/toolbars.py:
```diff
@@ -0,0 +1,249 @@
+        Override Qt method.
+
+        Take into account the existence of a corner widget when adding a new
+        widget in this toolbar.
```
```suggestion
        widget to this toolbar.
```

ccordoba12 commented on spyder/api/widgets/toolbars.py:
```diff
@@ -0,0 +1,249 @@
+        else:
+            super().addWidget(widget)
+
+    def addAction(self, action):
+        """
+        Override Qt method.
+
+        Take into account the existence of a corner widget when adding a new
+        action in this toolbar.
+        """
+        pass
```
Please remove this method if it's unnecessary. Or clarify why you're redefining it to make it pass in its docstring.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/api/widgets/auxiliary_widgets.py:
```diff
@@ -0,0 +1,82 @@
+        """
+        Override Qt method.
+        """
+        # TODO:
```
Do you plan to address this TODO before merging this?

goanpeca replied:
Fixed

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -699,16 +696,32 @@ class SpyderPluginV2(QObject, SpyderActionMixin, SpyderOptionMixin):
 
     # --- API: Signals -------------------------------------------------------
     # ------------------------------------------------------------------------
+    # Signals here are automatically connected by the Spyder main window and
+    # connected to the respective global actions defined on the main window
+    # widget
```
Correct

goanpeca commented on spyder/api/plugins.py:
```diff
@@ -1235,22 +1261,47 @@ def update_style(self):
-        menu = ApplicationMenu(self._main)
-        self._application_menus[name] = menu
+        self._main._APPLICATION_MENUS[name] = menu
+        # self.main.addToolBar(menu)
```
Fixed
